web/src/contexts/ConnectionContext.tsx

```diff
@@ -0,0 +1,151 @@
+          We're unable to reach the server right now. Please check that the 
+          installer is running and accessible.
+        </p>
+        
+        <div className="flex items-center justify-between">
+          <div className="flex items-center text-sm font-semibold text-gray-600">
+            <div className="animate-spin rounded-full h-4 w-4 border-b-2 border-blue-600 mr-2"></div>
+            Trying again in {Math.max(1, 10 - (retryCount % 10))} second{Math.max(1, 10 - (retryCount % 10)) !== 1 ? 's' : ''}
+          </div>
+          <button 
+            onClick={onRetry}
+            disabled={isRetrying}
+            className="px-4 py-2 bg-blue-600 text-white rounded hover:bg-blue-700 transition-colors disabled:opacity-50 disabled:cursor-not-allowed"
+          >
+            {isRetrying ? 'Retrying...' : 'Try Now'}
+          </button>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export const ConnectionProvider: React.FC<ConnectionProviderProps> = ({ children }) => {
+  const [isConnected, setIsConnected] = useState(true);
+  const [isChecking, setIsChecking] = useState(false);
+
+  const checkConnection = async () => {
+    setIsChecking(true);
+    
+    try {
+      // Try up to 3 times before marking as disconnected
+      let attempts = 0;
+      const maxAttempts = 3;
+      
+      while (attempts < maxAttempts) {
+        try {
+          // Create a timeout promise
+          const timeoutPromise = new Promise((_, reject) => 
+            setTimeout(() => reject(new Error('Timeout')), 1000)
+          );
+          
+          const fetchPromise = fetch('/api/health', {
+            method: 'GET',
+            headers: { 'Content-Type': 'application/json' },
+          });
+          
+          const response = await Promise.race([fetchPromise, timeoutPromise]) as Response;
+          
+          if (response.ok) {
+            setIsConnected(true);
+            return;
+          } else {
+            throw new Error(`HTTP ${response.status}`);
+          }
+        } catch {
+          attempts++;
+          if (attempts < maxAttempts) {
+            await new Promise(resolve => setTimeout(resolve, 500));
+          }
+        }
+      }
+      
+      // All attempts failed
+      setIsConnected(false);
+    } catch {
+      setIsConnected(false);
+    } finally {
+      setIsChecking(false);
+    }
+  };
+
+  useEffect(() => {
+    // Initial check
+    checkConnection();
+    
+    // Set up periodic health checks every 10 seconds
+    const interval = setInterval(checkConnection, 10000);
+    
+    return () => clearInterval(interval);
+  }, []);
+
+  const value: ConnectionContextType = {
+    isConnected,
+    isChecking,
+  };
+
+  return (
+    <ConnectionContext.Provider value={value}>
+      {children}
+      <ConnectionModal 
+        isVisible={!isConnected} 
+        onRetry={checkConnection}
+        isRetrying={isChecking}
+      />
+    </ConnectionContext.Provider>
```

Comment:
I wonder if we actually need this to be a context or if this could just be a stateless component handling props only instead 🤔

My understanding of context is it allows us to share state between deeply nested components without the need to pass props around. This would make sense for this case if we ever wanted other components to be aware of `ConnectionContextType` and check `isConnected` or `isChecking`, which I don't see happening?

Thoughts?

Reply:
Makes sense. I think when I started the idea was going to be to maybe present this modal more broadly on other components for other types of API calls as well, but decided to just use the the single `api/health` endpoint and keep things pretty simple. So not using react context i think is the right choice for now.